File: task4/task4.cpp
#include <cstdlib>
#include <ctime>
#include <vector>
#include <string>
#include <iostream>

using namespace std;

bool IsPrime(long long n) { //определение простое число или нет
    if (n <= 1) return false;
    for (long long i = 2; i * i <= n; i++) {
        if (n % i == 0) return false;
    }
    return true;
}

long long InDegreeByMod(long long a, long long b, long long m) { //возведение в степень по модулю
    long long result = 1;
    a = a % m;
    while (b > 0) { //условие завершения
        if (b % 2 == 1) result = (result * a) % m;
        b = b >> 1; //движемся вправо по биту, чтобы уменьшить число вдвое. Это равносильно делению, но из-за битовой операции будет быстрее
        a = (a * a) % m;
    }
    return result;
}

long long FindG(long long p) { //поиск примитивного корня по простому числу
    if (!IsPrime(p)) {
        cout << "Число p не является простым!" << endl;
        return -1;
    }

    srand(time(NULL)); //объявляем генератор
    long long gRand = rand() % (p - 1) + 1; //генерируем случайное число в диапазоне от 1 до (p - 1)число

    while (true) { //поиск примитивного корня
        bool primSqrt = true; //флаг - примитивный корень
        for (long long i = 2; i < p - 1; i++) { //возведение в степень по модулю
            if (InDegreeByMod(gRand, i, p) == 1) { //если мы получаем 1, то число не является примитивным корнем
                primSqrt = false;
                break;
            }
        }

        if (primSqrt) return gRand;

        gRand = rand() % (p - 1) + 1; //генерация нового числа
    }

    return -1;
}

long long GenSecKey(long long n) { //генерируем секретный ключ
    srand(time(NULL));
    return rand() % (n - 1) + 1;  //+1 используется для избигания ошибки
}

pair<long long, long long> GeneratePublicKey(long long p, long long g, long long x) {
    long long y = InDegreeByMod(g, x, p); //возводим в степень по модулю
    return make_pair(p, y); // из полученного значения создаётся пара чисел - открытый ключ
}

vector<pair<long long, long long>> EncryptString(const string &s, long long p, long long g, long long y) {
    vector<pair<long long, long long>> encrypted; //хранение зашифрованных чисел
    srand(time(NULL));
    for (char m : s) { //RBL
        long long k = rand() % (p - 1) + 1;
        long long c1 = InDegreeByMod(g, k, p); //(g^k) mod p
        long long c2 = (m * InDegreeByMod(y, k, p)) % p; //(m * y^k) mod p
        encrypted.push_back(make_pair(c1, c2)); //вносим шифрпару в вектор
    }
    return encrypted;
}

string DecryptString(const vector<pair<long long, long long>>& encrypted, long long p, long long x) { //расшифровка (вектор не должен менять значение - ссылка)
    string decrypted;
    for (auto& pair : encrypted) { //RBL
        long long c1 = pair.first;
        long long c2 = pair.second;
        long long m = (c2 * InDegreeByMod(c1, p - 1 - x, p)) % p; //(c2 * c1^(p-1-x)) mod p
        decrypted.push_back(static_cast<char>(m)); //m - расшифрованный символ сообщения; необходимо преобразовать m в char, чтобы в дальнейшем получить буквенный формат
    }
    return decrypted;
}

int main() {
    setlocale(LC_ALL, "Russian");

    cout << "Выберите действие:" << endl;
    cout << "1. Зашифровать сообщение" << endl;
    cout << "2. Расшифровать сообщение" << endl;
    cout << "3. Обмен" << endl;
    cout << "Ваш выбор: ";

    int choice;
    cin >> choice;

    long long p, g, x;
    pair<long long, long long> publicKey;

    switch (choice) {
        case 1: {
            cout << "Введите открытый ключ (p, g, y): ";
            cin >> p >> g >> x;
            cout << "Введите сообщение m для шифрования: ";
            string m;
            cin >> m;

            vector<pair<long long, long long>> txtEncrypt = EncryptString(m, p, g, x);
            cout << "Шифртекст: ";
            for (auto& pair : txtEncrypt) {
                cout << "(" << pair.first << ", " << pair.second << ") ";
            }
            cout << endl;
            break;
        }
        case 2: {
            cout << "Введите закрытый ключ x: ";
            cin >> x;
            cout << "Введите простое число p: ";
            cin >> p;
            cout << "Введите примитивный корень g: ";
            cin >> g;

            cout << "Введите количество пар шифртекста для расшифровки: ";
            int num_pairs;
            cin >> num_pairs;

            cout << "Введите шифртекст: ";
            vector<pair<long long, long long>> encryptedText;
            long long c1, c2;
            for (int i = 0; i < num_pairs; ++i) {
                cin >> c1 >> c2;
                encryptedText.push_back(make_pair(c1, c2));
            }

            string decrMsg = DecryptString(encryptedText, p, x);
            cout << "Расшифрованное сообщение: " << decrMsg << endl;
            break;
        }
        case 3: {
            cout << "Введите простое число p: ";
            cin >> p;

            g = FindG(p);
            if (g == -1) {
                cout << "Не найдено примитивного корня!" << endl;
                return 0;
            }

            cout << "Найденный примитивный корень g: " << g << endl;

            x = GenSecKey(p - 1);
            cout << "Секретный ключ x: " << x << endl;

            publicKey = GeneratePublicKey(p, g, x);
            cout << "Открытый ключ: (" << publicKey.first << ", " << g << ", " << publicKey.second << ")" << endl;
            break;
        }
        default: {
            cout << "Неверный выбор! Пожалуйста, выберите 1, 2 или 3." << endl;
        }
    }
    return 0;
}

File: task4/task4.rs
use rand::Rng;
use std::io;

fn is_prime(n: u64) -> bool {
    if n <= 1 {
        return false;
    }
    for i in 2..=(n as f64).sqrt() as u64 {
        if n % i == 0 {
            return false;
        }
    }
    true
}

fn in_degree_by_mod(mut a: u64, mut b: u64, m: u64) -> u64 {
    let mut result = 1;
    a %= m;
    while b > 0 {
        if b % 2 == 1 {
            result = (result * a) % m;
        }
        b >>= 1;
        a = (a * a) % m;
    }
    result
}

fn find_g(p: u64) -> u64 {
    if !is_prime(p) {
        println!("Число p не является простым!");
        return 0;
    }

    let mut rng = rand::thread_rng();
    let mut g_rand = rng.gen_range(1..p);

    loop {
        let mut prim_sqrt = true;

        for i in 2..p - 1 {
            if in_degree_by_mod(g_rand, i, p) == 1 {
                prim_sqrt = false;
                break;
            }
        }

        if prim_sqrt {
            return g_rand;
        }

        g_rand = rng.gen_range(1..p);
    }
}

fn gen_sec_key(n: u64) -> u64 {
    let mut rng = rand::thread_rng();
    rng.gen_range(1..n)
}

fn generate_public_key(p: u64, g: u64, x: u64) -> (u64, u64) {
    let y = in_degree_by_mod(g, x, p);
    (p, y)
}

fn for_encrypt(m: u64, p: u64, g: u64, y: u64) -> (u64, u64) {
    let mut rng = rand::thread_rng();
    let k = rng.gen_range(1..p);
    let c1 = in_degree_by_mod(g, k, p);
    let c2 = (m * in_degree_by_mod(y, k, p)) % p;
    (c1, c2)
}

fn decrypt(txt_encrypt: (u64, u64), p: u64, x: u64) -> u64 {
    let c1 = txt_encrypt.0;
    let c2 = txt_encrypt.1;
    (c2 * in_degree_by_mod(c1, p - 1 - x, p)) % p
}

fn main() {
    println!("Выберите действие:");
    println!("1. Зашифровать сообщение");
    println!("2. Расшифровать сообщение");
    println!("Ваш выбор: ");

    let mut choice = String::new();
    io::stdin().read_line(&mut choice).expect("Не удалось прочитать строку");
    let choice: u32 = choice.trim().parse().expect("Введите число");

    if choice == 1 {
        println!("Введите простое число p: ");
        let mut p = String::new();
        io::stdin().read_line(&mut p).expect("Не удалось прочитать строку");
        let p: u64 = p.trim().parse().expect("Введите число");

        let g = find_g(p);
        if g == 0 {
            println!("Не найдено примитивного корня!");
            return;
        }

        println!("Найденный примитивный корень g: {}", g);

        let x = gen_sec_key(p - 1);
        println!("Секретный ключ x: {}", x);

        let public_key = generate_public_key(p, g, x);
        println!("Открытый ключ (p, g, y): ({}, {}, {})", public_key.0, g, public_key.1);

        println!("Введите сообщение m для шифрования: ");
        let mut m = String::new();
        io::stdin().read_line(&mut m).expect("Не удалось прочитать строку");
        let m: u64 = m.trim().parse().expect("Введите число");

        let txt_encrypt = for_encrypt(m, public_key.0, g, public_key.1);
        println!("Шифртекст (c1, c2): ({}, {})", txt_encrypt.0, txt_encrypt.1);
    } else if choice == 2 {
        println!("Введите закрытый ключ x: ");
        let mut x = String::new();
        io::stdin().read_line(&mut x).expect("Не удалось прочитать строку");
        let x: u64 = x.trim().parse().expect("Введите число");

        println!("Введите простое число p: ");
        let mut p = String::new();
        io::stdin().read_line(&mut p).expect("Не удалось прочитать строку");
        let p: u64 = p.trim().parse().expect("Введите число");

        println!("Введите примитивный корень g: ");
        let mut g = String::new();
        io::stdin().read_line(&mut g).expect("Не удалось прочитать строку");
        let g: u64 = g.trim().parse().expect("Введите число");

        println!("Введите шифртекст (c1, c2): ");
        let mut c1 = String::new();
        io::stdin().read_line(&mut c1).expect("Не удалось прочитать строку");
        let c1: u64 = c1.trim().parse().expect("Введите число");

        let mut c2 = String::new();
        io::stdin().read_line(&mut c2).expect("Не удалось прочитать строку");
        let c2: u64 = c2.trim().parse().expect("Введите число");

        let decr_msg = decrypt((c1, c2), p, x);
        println!("Расшифрованное сообщение: {}", decr_msg);
    } else {
        println!("Неверный выбор. Пожалуйста, выберите 1 или 2.");
    }
}